applications/ShapeOptimizationApplication/python_scripts/custom_math.py

```diff
@@ -0,0 +1,422 @@
+
+        # apply delta
+        x = Plus(x,ScalarVectorProduct(alpha,dx))
+        y = Plus(y,ScalarVectorProduct(alpha,dy))
+        l = Plus(l,ScalarVectorProduct(alpha,dl))
+        rd,rp = rdrp(x,y,l,A,b)
+        k = k+1
+
+        # Determine current error
+        error = Norm2(VertCat(rd,rp))
+
+    if k==max_itr:
+        raise RuntimeError("custom_math::quadprodg: Suboptimization for projection to halfspaces did not converge in the specified number of iterations")
+    else:
+        # Exit code of 0 if convergence is reached
+        exit_code = 0
+
+    return x, k, error, exit_code
+
+# ------------------------------------------------------------------------------
+def PerformBisectioning(func, a, b, target, tolerance, max_itr):
+
+    fa_value_0, fa_is_converged = func(a)
+    fb_value_0, fb_is_converged = func(b)
+
+    error = None
+    itr = 0
+
+    # Always safe last argument for which function converges and return this if no solution is found
+    if fa_is_converged:
+        last_allowed_function_argument = a
+    elif fb_is_converged:
+        last_allowed_function_argument = b
+
+    # Check special cases
+    if abs(fa_value_0-target) < tolerance:
+        res_function_argument = a
+        error = abs(fa_value_0-target)
+
+    elif abs(fb_value_0-target) < tolerance:
+        res_function_argument = b
+        error = abs(fb_value_0-target)
+
+    elif abs(fa_value_0-fb_value_0) < 1e-13:
+        print("Bisectioning intervall yiels to identical function values!")
+        res_function_argument = a
+        error = abs(fb_value_0-target)
+
+    elif (fa_value_0-target)*(fb_value_0-target)>0:
+        print("Bisectioning on function, that has no root in specified intervall!! Returning the argument which yiels a closer value to the target.")
+
+        if abs(fa_value_0-target) < abs(fb_value_0-target):
+            res_function_argument = a
+            error = fa_value_0-target
+        else:
+            res_function_argument = b
+            error = fb_value_0-target
+
+    # Perform bisectioning if no special case applies
+    else:
+        p = (a + b)/2
+
+        for itr in range(1,max_itr+1):
+            fa_value, fa_is_converged = func(a)
+            fp_value, fp_is_converged = func(p)
+
+            if (fa_value-target)*(fp_value-target) < 0:
+                b = p
+            else:
+                a = p
+
+            p = (a + b)/2
+
+            fp_value, fp_is_converged = func(p)
+            error = abs(fp_value-target)
+
+            if fp_is_converged:
+                last_allowed_function_argument = p
+
+            if itr == max_itr:
+                res_function_argument = last_allowed_function_argument
+                print("Bisectioning did not converge in the specified maximum number of iterations!")
+                break
+
+            elif error < tolerance:
+                res_function_argument = p
+                break
+
+    return res_function_argument, itr, error
+
+
+
+
```

Comment:
I have some more general comments about this file:

1. Why are you using so many abbreviations? Makes it hard to read 
2. Small or capital letters are not consistently used
3. Sometimes function arguments have leading underscores, sometimes not
4. I have the feeling that a Vector and Matrix class that inherit from python list would make some things here more readable

I did not have a look in detail at the algorithms yet.

Reply:
All the functions in the custom_math.py will be moved to C++ once the whole "Projector"-class in the trust region algorithm is moved to C++. With this, a lot of the functions will be deleted as by no means basic linear algebra shall be duplicated. So the issues you mentioned will be fixed, once the Projector goes to C++.